### PR TITLE
Allow one to bound the size of output shards when writing to files.

### LINKED ISSUE #22129: [Feature Request]: Allow one to bound the size of output shards when writing to files.
### What would you like to happen?

By default, Beam automatically chooses the number of shards to write, but there are cases where these shards are either significantly larger or smaller than desired. Users can manually set the number of shards, but this comes at a(n often significant) performance cost as it forces all data to be reshuffled, and also requires guessing at the total output size to determine the correct number of shards. 

While creating two many small shards may be difficult to rectify (distinct bundles cannot write to the same shard), writing multiple shards in a single bundle is a straightforward way to enforce an upper bound on shard sizes.

### Issue Priority

Priority: 2

### Issue Component

Component: io-py-common

### PR DESCRIPTION
**Pull request by robertwb**
This fixes #22129 by possibly writing multiple shards per bundle.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @abacn

File: sdks/python/apache_beam/io/filebasedsink.py [hunk 1]
```diff
@@ -68,6 +68,9 @@ def __init__(
       shard_name_template=None,
       mime_type='application/octet-stream',
       compression_type=CompressionTypes.AUTO,
+      *,
```
Comment (Abacn) [inline sdks/python/apache_beam/io/filebasedsink.py:71]: Do we have code style guide about the usage of asterisk in function parameters?

File: sdks/python/apache_beam/io/filebasedsink.py [hunk 2]
```diff
@@ -406,10 +417,33 @@ def __init__(self, sink, temp_shard_path):
...
+    if self.sink.max_bytes_per_shard:
+      return (
+          self.sink.byte_counter.bytes_written >= self.sink.max_bytes_per_shard)
 
   def close(self):
     self.sink.close(self.temp_handle)
     return self.temp_shard_path
+
+
+class _ByteCountingWriter:
```
[the remark this answers is not in the document]
Comment (Abacn) [reply sdks/python/apache_beam/io/filebasedsink.py:436]: Found that io.BufferedWriter.write returns the number of bytes written (https://docs.python.org/3.7/library/io.html#io.BufferedWriter.write) so bytes_written can be traced directly in FileBasedSinkWriter.write

File: sdks/python/apache_beam/io/filebasedsink.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/io/filebasedsink.py:71]: I don't think we have any guidance here (other than that which is generic to Python, which would indicate most of these arguments should be passed by keyword).

File: sdks/python/apache_beam/io/filebasedsink.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/filebasedsink.py:436]: Unfortunately io.BufferedWriter.write returns the number of bytes written _for that call_, not a running total.

File: sdks/python/apache_beam/io/iobase.py [hunk 3]
```diff
@@ -848,8 +848,12 @@ class Writer(object):
   See ``iobase.Sink`` for more detailed documentation about the process of
   writing to a sink.
   """
-  def write(self, value):
-    """Writes a value to the sink using the current writer."""
+  def write(self, value) -> Optional[bool]:
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/iobase.py:851]: It's backwards compatible, which is why I made it Optional. But I've moved to using at_capacity as suggested instead.

File: sdks/python/apache_beam/io/iobase.py [hunk 4]
```diff
@@ -1184,7 +1188,9 @@ def process(self, element, init_result):
     if self.writer is None:
       # We ignore UUID collisions here since they are extremely rare.
       self.writer = self.sink.open_writer(init_result, str(uuid.uuid4()))
-    self.writer.write(element)
+    if self.writer.write(element):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/iobase.py:1191]: Done.

File: sdks/python/apache_beam/io/filebasedsink.py [hunk 5]
```diff
@@ -406,10 +417,33 @@ def __init__(self, sink, temp_shard_path):
     self.sink = sink
     self.temp_shard_path = temp_shard_path
     self.temp_handle = self.sink.open(temp_shard_path)
+    self.num_records_written = 0
 
   def write(self, value):
     self.sink.write_record(self.temp_handle, value)
+    if self.sink.max_records_per_shard:
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/filebasedsink.py:424]: Done.

File: sdks/python/apache_beam/io/filebasedsink.py [hunk 6]
```diff
@@ -108,6 +111,8 @@ def __init__(
         shard_name_template)
     self.compression_type = compression_type
     self.mime_type = mime_type
+    self.max_records_per_shard = max_records_per_shard
+    self.max_bytes_per_shard = max_bytes_per_shard
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/filebasedsink.py:115]: Nice catch. Fixed so that both take effect.

**Review by robertwb — COMMENTED** [review]
Thanks. I've addressed your comments. We do have to push the change down to iobase as that's where the writers are created (and destroyed) but I think the change there should be minimal and natural.

**Review by Abacn — APPROVED** [review]
Thanks! LGTM

**Comment by robertwb** [comment]
Unrelated failure in apache_beam.runners.portability.fn_api_runner.fn_runner_test.FnApiRunnerTestWithGrpcAndMultiWorkers.test_pardo_large_input

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by lhoestq** [comment]
Hi ! is this available for `WriteToParquet` as well ? I couldn't find a way to do it with Parquet files

**Comment by robertwb** [comment]
This has not been plumbed through for `WriteToParquet`, but the underlying infrastructure should work there (see the changes to sdks/python/apache_beam/io/textio.py -- just adding arguments and passing them through). I'd be happy to help you with a pull request.

**Pull request merged**